packages/from-hardhat/src/compilation.ts

```diff
@@ -0,0 +1,121 @@
+import type * as Hardhat from "hardhat/types";
+
+import type { Compilation, CompiledContract } from "@truffle/compile-common";
+import * as CompileSolidity from "@truffle/compile-solidity";
+
+import { supportedHardhatBuildInfoFormats } from "./constants";
+
+export const fromBuildInfo = (buildInfo: Hardhat.BuildInfo): Compilation => {
+  const { _format } = buildInfo;
+
+  if (!supportedHardhatBuildInfoFormats.has(_format)) {
+    throw new Error(`Unsupported build info format: ${_format}`);
+  }
+
+  const sourceIndexes = SourceIndexes.fromBuildInfo(buildInfo);
+
+  return {
+    sourceIndexes,
+    sources: Sources.fromBuildInfo(buildInfo, sourceIndexes),
+    contracts: Contracts.fromBuildInfo(buildInfo),
+    compiler: {
+      name: "solc",
+      version: buildInfo.solcLongVersion
+    }
+  };
+};
+
+namespace SourceIndexes {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo
+  ): Compilation["sourceIndexes"] => {
+    const sourceIndexes = [];
+    for (const { index, sourcePath } of Object.entries(
+      buildInfo.output.sources
+    ).map(([sourcePath, source]) => ({ index: source.id, sourcePath }))) {
+      sourceIndexes[index] = sourcePath;
+    }
+
+    return sourceIndexes;
+  };
+}
+
+namespace Sources {
+  export const fromBuildInfo = (
+    buildInfo: Hardhat.BuildInfo,
+    sourceIndexes: Compilation["sourceIndexes"]
+  ): Compilation["sources"] =>
+    sourceIndexes.map(sourcePath => {
+      // to handle if sourceIndexes is a sparse array
+      if (!sourcePath) {
+        return;
+      }
```

Comment:
Doesn't this also return a sparse array with an `undefined` at this index? Maybe you want to remove sparse indices with a `.filter` ahead of the map? I think you could also avoid the extra `O(n)` pass of a `filter` line by using a `reduce` that only accumulates defined `sourcePath` values into an array.

Personally I tend to find the `filter` approach more readable, however.

Reply:
Actually this was probably just overly defensive programming. I realized afterwards that a lot of stuff would break if this was actually sparse.

Reply:
But it is more important to get `sourceIndexes` correct, and `sources` could be in any order

Reply:
I suppose it might be better in that case to just remove this block and the associated comment.